Add language selection dropdown

ignotus666 opened the pull request:
**Short description of changes**
Adds a language selection dropdown to the website. ~~The website also detects what language the user's browser is in and initially displays itself in that language~~.

Scripts automating the addition of new languages have been updated to support adding the full language name + language code.

It also includes the changes from https://github.com/jamulussoftware/jamuluswebsite/pull/988 and https://github.com/jamulussoftware/jamuluswebsite/pull/990 so as to avoid conflicts between PRs and include all the changes here.

TODO after merging this:
- Change language codes in po folder names (in po files too? or on Weblate?).
- Merge `add-lang` and `po4a-add-language` scripts into release.
See #993.

**Context: Fixes an issue? Related issues**
Fixes https://github.com/jamulussoftware/jamuluswebsite/issues/977, https://github.com/jamulussoftware/jamuluswebsite/issues/986 and https://github.com/jamulussoftware/jamuluswebsite/issues/992

**Status of this Pull Request**
Should be ready pending a review.

**What is missing until this pull request can be merged?**
Would be good if some more people can test it.

IMPORTANT: Weblate stuff should be merged first.

**Does this need translation?**

No.

ignotus666 committed:
Lang select (#11)

* Initial replacements testing OK so far

* Add language selection dropdown

* Don't store lang preference

* Align dropdown to the right. Remove unneeded code from fw.css

* Make dropdown scrollable

* Clean/move/improve css styling, add icon, lang sel. outline

* Remove unneeded css

* Show full lang names, automate adding langs

* AUTO: Add new language: (eu) #4

* Revert "AUTO: Add new language: (eu) #4"

This reverts commit 301c9486933621f8b5ae9300b4b7ba106b55662e.

* Correction

* AUTO: Add new language: (Basque) #5

* Update add-lang.yml

* AUTO: Add new language: (English (American)) #6

* Correction

* Revert language additions

* Correction

* Correction

* Update

* Update

* Add check

* Update error message

* Revert lang code hyphens

* Add changes by @gilgongo

* Combine with gilgongo's stuff

* Correct title check

* Correct pt-BR language codes

* Revert lang code changes

---------

Co-authored-by: Gilgongo <gilgongo@phreak.co.uk>
Co-authored-by: github-actions <41898282+github-actions[bot]@users.noreply.github.com>

ignotus666 commented:
As commented in #993, the check fails because of lang code mismatches that will be fixed after this PR is merged. To test the language dropdown menu, please use the artifact from [here](https://github.com/jamulussoftware/jamuluswebsite/actions/runs/9616653685).

ann0see commented on _includes/langselect.html:
```diff
@@ -1,17 +1,218 @@
+  });
+</script>
+<style>
+#langselect-container {
```
Please put this into a separate css file. This should NOT be in the document!

ignotus666 replied:
In a file of its own with just this styling section, or what? All the stying?

ann0see replied:
Yes. You should always put CSS into a separate file as it can be cached.

ignotus666 replied:
Ok, so does this separate file need to have a specific name or is just putting it in `assets/css/` enough?

ignotus666 replied:
I think I answered myself - I put it all in `fx.css` and it seems happy enough. Would that do?

ann0see replied:
langsel.css

ann0see replied:
Changed my mind to fw.css. See below

ann0see commented on _includes/langselect.html:
```diff
@@ -1,17 +1,218 @@
+  </div>
 </div>
+<script>
+  document.addEventListener('DOMContentLoaded', function() {
```
Initially I didn't implement something like this on purpose: users may want to switch with a more conscious choice and google should redirect to the correct language by default if the meta tags are set correctly.

Anyway: The code looks a bit too large and complex for a feature like that. I'll have a look at this some time.

pljones replied:
Also, same as CSS -- we shouldn't put javascript inline in the HTML files.

ann0see commented on _includes/langselect.html:
```diff
@@ -1,17 +1,218 @@
+    var defaultLang = '{{ site.default_lang }}';
+
+    // Get the stored language from session storage
+    var storedLang = sessionStorage.getItem('activeLang');
```
Storing data on the client is something which I wanted to avoid initially. One goal of the site was to be as "dumb" as possible and work 99% without JS. But we can of course change this.

ignotus666 replied:
I've tried to get language detection to work without any kind of storage but then it does weird things like keep redirecting to the default/browser language.

ann0see replied:
Yes. This is expected. If we have an auto redirect, it will always redirect if we don't store the state. I'm not too much a fan of auto redirecting - rather have a small message which the user can then select instead of doing too much automagically. Check how https://apple.com/it just shows a message right at the top.

ignotus666 replied:
My main concern was getting rid of the ugly string of lang codes for language selection and tidying it up with a dropdown menu. Then it kind of blew up into the auto-redirect version after a request by @henkdegroot. If that involves implementing stuff we'd rather not have, I really don't care either way - not having it would simplify things significantly.

I've cobbled it together with a tenuous understanding of what I'm doing, to put it mildly... So AFAIC, the simpler the better.

ann0see commented on _includes/langselect.html:
```diff
@@ -1,17 +1,218 @@
+
+    if (shouldRedirect) {
+      var currentUrl = window.location.pathname;
+      var newUrl = newLang === defaultLang ? currentUrl.replace(/^\/[^/]+/, '') : '/' + newLang + currentUrl.replace(/^\/[^/]+/, '');
```
There must be an easier way. Get the content of the respective language of the language picker? Then this doesn't need to be duplicated here.

ann0see commented on _includes/langselect.html:
```diff
@@ -1,17 +1,218 @@
+      var currentUrl = window.location.pathname;
+      var newUrl = newLang === defaultLang ? currentUrl.replace(/^\/[^/]+/, '') : '/' + newLang + currentUrl.replace(/^\/[^/]+/, '');
+
+      // Prevent endless redirect loops by checking if the URL already contains the language
```
Maybe we can already check this earlier.

ann0see commented on _includes/langselect.html:
```diff
@@ -1,17 +1,218 @@
+      var newUrl = newLang === defaultLang ? currentUrl.replace(/^\/[^/]+/, '') : '/' + newLang + currentUrl.replace(/^\/[^/]+/, '');
+
+      // Prevent endless redirect loops by checking if the URL already contains the language
+      if (!window.location.pathname.startsWith('/' + newLang) && window.location.pathname !== newUrl) {
```
checking for absolute paths "/" may need some more thought. I'd like to get most of the stuff from the language picker.

ann0see reviewed (commented):
In general, I see the benefits. The redirect would be a new feature and it seems to be possible to click the picker to get back to another language if it fails.

ann0see commented on .github/workflows/add-lang.yml:
```diff
@@ -34,11 +36,11 @@ jobs:
             exit 1
           else
-            echo Language: [${{ env.new_lang }}] not detected. Proceeding to add [${{ env.new_lang }}].
+            echo Language: [${{ env.lang_code }}] not detected. Proceeding to add [${{ env.lang_code }}].
```
This should be made a bit more explanatory: Language: ... not detected doesn't explain really what this means. Maybe say something like language ... seems to be new as no Language: ... was found. Proceeding to add...

ignotus666 replied:
Done.

ann0see commented on _includes/langselect.html:
```diff
@@ -1,17 +1,218 @@
-<div id="langselect">
-  <div lang="en" id="suglang">This site might also be available in your native language. Select your preferred language below:</div>
```
Please check if there's still JS which uses the suglang id

ann0see commented on _includes/langselect.html:
```diff
@@ -1,17 +1,218 @@
+    <div class="language-dropdown">
+      <button class="dropdown-button">
+        <span class="dropdown-button-content">
+          <img src="/assets/img/lang-icon.png" alt="Language Icon" class="lang-icon">
```
Don't use absolute links here. Jekyll should support building the path somewhat dynamically.

ann0see commented on assets/img/lang-icon.png:
Where does this icon come from? Can we also have a .svg instead?

ignotus666 replied:
I took it from a site with open-licenced icons and changed the colours. I have a .svg that can replace the .png.

ann0see commented on sitemap.xml:
```diff
@@ -31,7 +31,7 @@ category: hidden
     {% unless post.published == false %}
       <url>
-        <loc>{{ site.url }}{{ post.url }}</loc>
+        <loc>{{ site.a_rootpage }}{{ post.url }}</loc>
```
What does this do?

ignotus666 replied:
That's from #988 - @gilgongo should know.

gilgongo replied:
@ann0see  That was to fix the fact that we weren't publishing fully-qualified paths for the KB URLs in sitemap.xml. Google complained they were invalid so wouldn't index them.

ann0see replied:
Ok.

ann0see commented:
What would a classic html select/option box with some styling be as alternative? It would be much better integrated in mobile OS and be standard compliant: 
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select?retiredLocale=de

ann0see commented:
I'd like to be part of the final review on this one.

pljones commented:
Regarding avoidance of reliance on javascript:
- have the default site written with the "Language Select" as a standard popup that has a link for each language that takes you to the current url in that language
- have an onLoad script that replaces that with the "Language Select" widget and also runs the auto-detect stuff

That should be pretty transparent to the user - the benefit of the auto-detect (and then the widget having local storage) would be there for those who are happy to have javascript and local storage enabled but the functionality wouldn't be lost for other users.

ignotus666 commented:
@ann0see, @pljones: I have a new version that doesn't do language detection nor uses javascript, and the css styling is moved to its own `langsel.css` file. The code is way more simplified. Other than not detecting the language automatically, it behaves exactly the same. So subject of course to a review, would you be happier if I replaced the current version with that?

ann0see commented:
Yes.

ignotus666 committed:
Replace with simplified version

gilgongo commented:
We'll need to verify the accessibility of this at some point too. Not sure how best to do that though.

ann0see commented:
As I said above, I'd assume that a select box is more accessible. But IDK if clicking easily works without JS. We could use a form.

ignotus666 commented:
[Here's](https://github.com/ignotus666/jamuluswebsite/actions/runs/9675967431) an artifact you can test.

ignotus666 commented:
The last update should address accessibility.

ignotus666 committed:
Change add-lang message, remove suglang reference

ignotus666 committed:
Improve accessibility

ignotus666 commented:
The latest version should address accessibility. Artifact [here](https://github.com/ignotus666/jamuluswebsite/actions/runs/9691999145).

ignotus666 committed:
Match text colour

ignotus666 committed:
Make dropdown menu more compact

ann0see commented on _includes/general/nav.html:
```diff
@@ -30,6 +33,6 @@ <h3>{{ item.title }}</h3>
       {% endif %}
     </ul>
-	<div id="cc">{{ site.data.copyright.nav.docs }}</div>
+        <div id="cc">{{ site.data.copyright.nav.docs }}</div>
```
Formatting seems strange

ann0see replied:
@ignotus666 please check if formatting on line 36 is ok.

pljones replied:
Looks like a tab changed to spaces.

ignotus666 replied:
This file has been totally reformatted since the comment was made.

pljones replied:
So this can be resolved?

ignotus666 replied:
If everyone's happy with the current formatting, then I suppose so

ann0see commented on _includes/langselect.html:
```diff
@@ -1,17 +1,37 @@
-      {% endif %}
-    {% endfor %}
-  </ul>
+<div id="langselect-container">
```
Please check if the naming is consistent with ids found elsewhere

ann0see replied:
Closing as style only.

ann0see commented on _includes/langselect.html:
```diff
@@ -1,17 +1,37 @@
+        <span class="dropdown-button-content">
+          <img src="{{ site.baseurl }}/assets/img/lang-icon.svg" alt="Language Icon" class="lang-icon">
+          <span class="lang-text">{{ site.language_names[site.active_lang] }} &#9660;</span>
+          <link rel="stylesheet" href="{{ site.baseurl }}/assets/css/langsel.css">
```
No. This MUST be in the head tag

ann0see commented on _includes/langselect.html:
```diff
@@ -1,17 +1,37 @@
+          <link rel="stylesheet" href="{{ site.baseurl }}/assets/css/langsel.css">
+        </span>
+      </button>
+      <div class="dropdown-content" id="dropdown-menu" role="menu">
```
I still don't believe in the manual approach. Usually, I'd expect a dropdown box from HTML. This follows system standards on mobile and is probably the most accessible

ann0see commented on _includes/langselect.html:
```diff
@@ -1,17 +1,37 @@
+            {% else %}
+              {% if lng == site.default_lang %}
+                {% capture langlink %}
+                  {{ site.baseurl }}{{ page.url | replace: '/*' }}
```
Seems odd to use replace here.

ann0see commented on assets/css/langsel.css:
```diff
@@ -0,0 +1,118 @@
+#langselect-container {
```
Maybe we'd just put this in fw.css as we need to load this on every page.

ignotus666 replied:
Done.

ann0see commented:
You mentioned somewhere that you used ChatGPT. Does this have any implications?

ignotus666 commented:
I used Chatgpt to give me pointers when stuck (admittedly fairly often). I doubt this has any implications at all.

> I still don't believe in the manual approach. Usually, I'd expect a dropdown box from HTML. This follows system standards on mobile and is probably the most accessible

If anything other than an entirely different approach is ultimately going to be rejected, I'd rather stop wasting my time instead of rearranging the deckchairs on this Titanic of a PR.

ignotus666 committed:
Make dropdown standard html

ignotus666 commented:
Last try. It now uses standard HTML and is much simpler. I couldn't get the icon to integrate in the button without things messing up so I got rid of it. [Artifact](https://github.com/ignotus666/jamuluswebsite/actions/runs/9740714163).

![lang-select_1](https://github.com/jamulussoftware/jamuluswebsite/assets/4263412/3462b996-b85e-40b8-b0dd-c2d6356351f1)

ignotus666 committed:
Replace with simplified version

Change add-lang message, remove suglang reference

Improve accessibility

Match text colour

Make dropdown menu more compact

Make dropdown standard html

ignotus666 committed:
Merge branch 'lang-select2' of https://github.com/ignotus666/jamuluswebsite into lang-select2

ann0see commented on _includes/langselect.html:
```diff
@@ -2,36 +2,21 @@
-          {% endfor %}
-        </ul>
-      </div>
+      <select id="language-select" onchange="location = this.value;">
```
Exactly. The onchange may be JS though.

ann0see replied:
Yes. It is. So we should have an alternative for non JS. Add a plain list (like before) and only show it on noscript.
Hide it with langsel.css but force show it in noscript.css
Or wrap it in a `<form>`. I believe you need to in any case.

ignotus666 replied:
Done.

ignotus666 committed:
Add no-JS option

ignotus666 commented:
Added non-JS option. [Artifact](https://github.com/ignotus666/jamuluswebsite/actions/runs/9756529488).

ignotus666 commented on _includes/langselect.html:
```diff
@@ -1,17 +1,39 @@
+            {% capture langlink %}{{ site.baseurl }}/{{ lng }}{{ page.url }}{% endcapture %}
+            {% capture lang_rel %}alternate{% endcapture %}
+          {% endif %}
+          <li><a href="{{ langlink }}">{{ site.language_names[lng] }}</a></li>
```
@gilgongo: not sure if something needs to be added here for the canonical links. And also a bit further up, line 16.

ignotus666 committed:
Correction to no-JS list

ignotus666 commented on _includes/langselect.html:
```diff
@@ -1,17 +1,39 @@
+              {% capture langlink %}{{ site.baseurl }}/{{ lng }}{{ page.url }}{% endcapture %}
+              {% capture lang_rel %}alternate{% endcapture %}
+            {% endif %}
+            <option value="{{ langlink }}" {% if lng == site.active_lang %}selected{% endif %}>
```
@gilgongo I'm not sure if/how something has to be added here for the canonical links.

gilgongo replied:
As long as the markup shows any link that goes to a `site.active_lang` page as `canonical` and (I think less importantly) any other langs as `alternate`, we're OK.

gilgongo replied:
Whoops, @ignotus666 sorry I meant `site.default_lang` (if that's a thing?)

gilgongo replied:
Only just got round to looking properly.

The markup seems to be coming out as `lang="canonical"` and `lang="alternate"` here when I view it with Jekyll. They need to be `hreflang="[lang]"` with the actual lang code of the URL. 

When the currently selected language isn't `site.default_lang`, URLs need to have `rel="alternate"`, otherwise they are `rel="canonical"` (which I think is happening?)

ignotus666 replied:
@gilgongo did you test what happens with JS disabled? Does it work as intended then?

gilgongo replied:
Most of it seems OK with and without JS (eg at the top of the page) but some links like this in the nav (I think - hard to tell) are showing`<ul><li><a href="/wiki/Getting-Started" lang="canonical" class="">Setting Up</a><ul>` when they should be `rel="canonical"` and have `hreflang` tags.

ignotus666 replied:
Could you test [this](https://github.com/ignotus666/jamuluswebsite/actions/runs/9875448341) version please? Sorry, I don't know what I should be looking for...

gilgongo replied:
That version (I've managed to un-minify it now) is almost OK but the nav seems to have slipped up a bit. It should be ` hreflang="es" ` here (not `lang="alternate"`) with `rel="alternate"` added. 

```
<h3>Cómo usar Jamulus</h3>
  <ul>
   <li>
   <a href="/es/wiki/Getting-Started" lang="alternate" class="">Configuración</a>
   <ul>
   <li><a href="/es/wiki/Installation-for-Windows" lang="alternate" class="">Windows</a></li>
   <li><a href="/es/wiki/Installation-for-Macintosh" lang="alternate" class="">macOS</a></li>

```

Otherwise I think all the other links are OK (eg the ones in the `<head>` section). Essentially, en needs to **always** have `rel="canonical"`, since that (I hope!) will fix the issue of Google being confused when it finds the same URL given as a different language that has not (or has only partially) been translated. After that, all other languages need to identify their links with `rel="alternate"`.

BTW seems to be a glitch in the selector's `<noscript>` section maybe? For example ES is identifying as EN here:

```
<noscript>
    <ul class="language-list">
    <li><a href="/de/" rel="alternate" hreflang="de" class=""> de </a></li>
    <li><a href="/es/" rel="canonical" hreflang="en" class=""> en </a></li>
    <li><a href="/es/" rel="alternate" hreflang="es" class="active"> es </a></li>
    <li><a href="/fr/" rel="alternate" hreflang="fr" class=""> fr </a></li>
```

ignotus666 replied:
@gilgongo according to [this](https://developers.google.com/search/docs/crawling-indexing/consolidate-duplicate-urls), the canonical link reference should be in the `head` section, which in our case is in `mainhomepage.html`. I tried adding the canonical stuff to `langselect.html` with an include in `mainhomepage.html` to avoid cluttering it up but that resulted in a duplicate dropdown button at the top of the page. So I created a separate `lang-links.html` file and that seems to work.

Could you please test [this](https://github.com/ignotus666/jamuluswebsite/actions/runs/9935684488) version?

ignotus666 replied:
> BTW seems to be a glitch in the selector's <noscript> section maybe? For example ES is identifying as EN here:

That happens with whichever the active language is - my guess is that it's supposed to be that way.

gilgongo replied:
> @gilgongo according to [this](https://developers.google.com/search/docs/crawling-indexing/consolidate-duplicate-urls), the canonical link reference should be in the `head` section, which in our case is in `mainhomepage.html`. 

Ah OK, ignore my comment about the `rel=` tag on URLs outside of the head section then - I didn't realise that wasn't needed.

`lang="canonical"` is still in the URLs in the main nav section though - should that not have the language code in it? That is (for English) `lang="en"` ?

ann0see replied:
> That happens with whichever the active language is - my guess is that it's supposed to be that way.

I don't think so. I'd consider this to be a bug.

ann0see replied:
Also, in the current version you cannot click the language you've currently selected - this makes sense for the non JS version.

ignotus666 replied:
@ann0see Ok, it's simple enough for non-JS mode, but seems to be tricky for JS mode without a script. Is JS mode good as it is or do we also want the currently selected language to be non-clickable in the dropdown?

ann0see replied:
Yes. I think it's only needed for non JS mode

ignotus666 replied:
Ok done. Artifact is at the bottom.

ignotus666 replied:
> I don't think so. I'd consider this to be a bug.

@ann0see I can't for the life of me figure out how to stop it doing that. Nothing I try changes that behaviour.

ignotus666 replied:
I think this is fixed now.

ignotus666 committed:
Make non-JS selection highlighted

ignotus666 committed:
Add language-container to fw.css

ignotus666 committed:
Correct translation readme

gilgongo commented:
@ignotus666 I'm not able to run this locally as I'm getting `No such file or directory`  on the `1-[lang]-index.md` files when I run `jekyll serve`. Do I need to tweak something?

ignotus666 commented:
@gilgongo are you using the artifact I linked to above? A clone of this branch won't work.

gilgongo commented:
> @gilgongo are you using the artifact I linked to above? A clone of this branch won't work.

Oh OK. How do I view it with jekyll?

ignotus666 commented:
Unzip it (twice; a second zipped folder emerges from the first). You should then have a folder called `_site.` CD into that and then use the `jekyll serve` command.

ann0see commented:
As validity check, upload the generated HTML to https://validator.w3.org/#validate_by_upload

pljones commented:
@jamulussoftware/designers @jamulussoftware/translators @jamulussoftware/maindevelopers 
Are we okay to merge this?  If so, please approve.

trebmuh reviewed (approved):
:+1:

gilgongo commented:
> @jamulussoftware/designers @jamulussoftware/translators @jamulussoftware/maindevelopers Are we okay to merge this? If so, please approve.

Just need to fix the html tags to make it do the stuff in I think #988

ann0see commented:
Yes. And I need to check it locally.

ignotus666 committed:
Fix canonical/alternate links

ignotus666 commented:
@ann0see the latest version for checking locally is [here](https://github.com/ignotus666/jamuluswebsite/actions/runs/9935684488).

softins commented:
> @ann0see the latest version for checking locally is [here](https://github.com/ignotus666/jamuluswebsite/actions/runs/9935684488).

I downloaded the artifact and unzipped it, but it seems rather tricky to check it locally without setting up a local web server with its own local domain to put the site at the root of a domain. There are a LOT of links in both `<head>` and `<body>` that are absolute instead of relative, either beginning with `/` or even with `https://jamulus.io/`, which means it can't be viewed properly from a sub-folder.

How do others check the `next-release` branch or PRs for it?

gilgongo commented:
@softins It's  bit convoluted but once you've unzipped it (twice) you can view it locally by building it with jekyll. I [outlined the process here](https://github.com/jamulussoftware/jamuluswebsite?tab=readme-ov-file#viewing-the-website-offline) but YMMV.

softins commented:
Thanks, I'll give it a go.

ann0see commented:
![language duplication](https://github.com/user-attachments/assets/8c7b467e-0f9d-49fb-b595-149805f4669c)

The selection box shows up twice. And I don't think it looks nice in the footer.

ann0see commented:
@ignotus666 also please check with https://github.com/jamulussoftware/jamuluswebsite/pull/994#issuecomment-2223833457 multiple issues are showing up in the w3c validator.

ignotus666 commented:
> The selection box shows up twice. And I don't think it looks nice in the footer.

Oops, a stray line got left in there (but not in this PR). Fixed version is [here](https://github.com/ignotus666/jamuluswebsite/actions/runs/9951953716).

> also please check with https://github.com/jamulussoftware/jamuluswebsite/pull/994#issuecomment-2223833457 multiple issues are showing up in the w3c validator.

What file am I supposed to upload there, `langselect.html`? It also spits out a bunch of warnings and errors with the current (`next-release`) version.

ann0see commented:
Ok. No you should upload the index.html and check for new errors then.

gilgongo commented:
> What file am I supposed to upload there, `langselect.html`? It also spits out a bunch of warnings and errors with the current (`next-release`) version.

I uploaded a jekyll-generated version of the home page.

ignotus666 commented:
Right. So the main complaint (that differs from the next-release errors) seems to be that "canonical" in `lang="canonical"` is more than 8 characters long... How are we supposed to get around that?

gilgongo commented:
My understanding is that `lang=` needs to be `lang="[langcode]"` (eg `lang="ko-KR"`) and that for links outside of the `<head>` section we don't need to have `rel="canonical"` and `rel="alternate"` as you point out (but could have `rel="[something else]"` if needed I guess)

ignotus666 commented:
I traced the problem back to nav.html. I think it's sorted now. Link to [artifact](https://github.com/ignotus666/jamuluswebsite/actions/runs/9954424944).

gilgongo commented:
That looks good. I don't know whether the links in the nav to things like `"/wiki/Getting-Started"` also need to have `hreflang=` in them but perhaps that's not needed with the `rel=` stuff in the head area.

ignotus666 commented:
Here's what I found: 

> Use rel="canonical" in navigation links for the primary version of a page.
> Reserve hreflang for <link> elements in the <head> to specify alternate language versions of pages.

ignotus666 committed:
Fix canonical/alternate links

ann0see commented on _layouts/mainhomepage.html:
```diff
@@ -30,6 +30,9 @@
   </script>
+
+  <!-- Include the lang-links.html for canonical and alternate links -->
+  {% include lang-links.html %}
```
Why just on the main homepage?

ignotus666 replied:
Because I probably mistakenly assumed that's where it should go. As everything seems to work, I assume it's ok. If it belongs somewhere else, just tell me.

ann0see replied:
I'd believe that the content in headtags.html (which gets included in every site) would suffice? Most likely, we'd have duplication here - but I'm not sure.

ignotus666 replied:
@ann0see If I replace the canonical/alternate code in `headtags.html `with the content of `lang-links.html` (and remove the latter), and also remove the include from `mainhomepage.html`, everything works. I guess this is preferable as it's much tidier. If that sounds good to you I'll update this PR with it.

ann0see replied:
Yes. And check if the content looks sensible in all pages.

ignotus666 replied:
@ann0see done. While testing I found that switching back to EN after choosing another language wasn't working (in non-JS mode)... that's fixed. Everything looks good clicking through all the pages in all languages. The link at the bottom has been changed to the current latest artifact.

ann0see commented on _po4a-tools/po4a-add-language.sh:
```diff
@@ -8,38 +8,40 @@ SCRIPT_DIR=$( cd "$(dirname "${BASH_SOURCE[0]}")" ; pwd -P )
 
 # Create and populate .po file folder
-mkdir ../_translator-files/po/$new_lang
+mkdir ../_translator-files/po/$lang_code
```
Having a bad feeling about not quoting this:
```suggestion
mkdir ../_translator-files/po/"$lang_code"
```

may work?

ann0see replied:
The style is different in the main repo though (and the main repo is probably safer). We're way more strict there. See e.g. https://github.com/jamulussoftware/jamulus/blob/main/.github/autobuild/linux_deb.sh

pljones replied:
Surely `$lang_code` is semantically and syntactically incorrect if it contains spaces, so the quotes are redundant?

ann0see replied:
Yes. The regex should not match in this case.

pljones replied:
```suggestion
mkdir "../_translator-files/po/${lang_code}"
```
would be more in keeping with the main repo, if we're going to refactor.

ignotus666 replied:
The style in this PR just follows that used in the rest of the po4a scripts. I think a clean-up of all the scripts should happen in a separate PR and do them all at once.

ann0see replied:
Ok. Let's leave this for another PR

ignotus666 committed:
Make current language non-selectable in non-JS mode

ignotus666 commented:
Latest version making selected language non-clickable in non-JS mode: [artifact](https://github.com/ignotus666/jamuluswebsite/actions/runs/10076102286).

ignotus666 committed:
Tidy up and fix non-selectable EN

ann0see commented on _includes/headtags.html:
```diff
@@ -1,15 +1,18 @@
+    <link rel="alternate" hreflang="{{ lng }}" href="{{ site.url }}{{ site.baseurl }}/{{ lng }}{{ page.url }}">
+  {% endif %}
 {% endfor %}
+<link rel="alternate" hreflang="x-default" href="{{ site.url }}/{{ site.default_lang }}{{ page.url }}">
```
I believe that this is not correct: it will produce /en/ but there is no /en/ subpage.

```suggestion
<link rel="alternate" hreflang="x-default" href="{{ site.url }}/{{ page.url }}">
```

is probably correct.

ann0see replied:
I hope the whole canonical/alternate thing is correct then.

gilgongo replied:
Oh, I see Google says "The new` x-default` hreflang attribute value signals to our algorithms that this page doesn't target any specific language or locale and is the default page when no other page is better suited."

ann0see replied:
Which we could argue is ok for en.

ignotus666 replied:
> I believe that this is not correct: it will produce /en/ but there is no /en/ subpage.

Yes, you're right.

ignotus666 committed:
Update _includes/headtags.html

Co-authored-by: ann0see <20726856+ann0see@users.noreply.github.com>

ann0see commented:
Ok. Thanks! It was a long ride but now hopefully it's all good. We'll probably need to check the site before publication for e.g broken links and so.

ignotus666 commented:
> It was a long ride

You can say that again... ;) . Right, now before unlocking Weblate or merging anything else here there are a couple of things that need doing (see description of this PR). Hopefully I'll have some time tonight to put it together.

gilgongo commented:
@ignotus666 I'm looking the the `rel=canonical/alternate` stuff and I'm not sure it's correct. Shall I open a PR with a couple of tweaks to `headtags.html`? If so, would that be against `main` or `next-release`? 

I also notice a `rel="canonical` in the `1-*-index.md` files which probably shouldn't be there - how best to remove that?

ann0see commented:
In theory next-release. But I'd rather wait until the translations have settled even though it shouldn't have conflicts
